File: AssetSync/src/AssetSync.Infrastructure/Api/WebhookService.cs
using System.Text;
using System.Text.Json;
using AssetSync.Core.Interfaces;
using AssetSync.Core.Models;

namespace AssetSync.Infrastructure.Api;

public class WebhookService : IWebhookService
{
    private readonly string? _webhookUrl;
    private readonly string _webhookType;
    private readonly IHttpClientFactory _httpClientFactory;

    public WebhookService(string? webhookUrl, string webhookType, IHttpClientFactory httpClientFactory)
    {
        _webhookUrl = webhookUrl;
        _webhookType = webhookType;
        _httpClientFactory = httpClientFactory;
    }

    public async Task SendSyncNotificationAsync(SyncRunSummary summary, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrEmpty(_webhookUrl)) return;
        var (content, contentType) = FormatSyncPayload(summary);
        var client = _httpClientFactory.CreateClient();
        var body = new StringContent(content, Encoding.UTF8, contentType);
        await client.PostAsync(_webhookUrl, body, cancellationToken).ConfigureAwait(false);
    }

    public async Task SendConnectivityFailureNotificationAsync(string serviceName, string message, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrEmpty(_webhookUrl)) return;
        var (content, contentType) = FormatAlertPayload("Connectivity failure", $"{serviceName}: {message}");
        var client = _httpClientFactory.CreateClient();
        var body = new StringContent(content, Encoding.UTF8, contentType);
        await client.PostAsync(_webhookUrl, body, cancellationToken).ConfigureAwait(false);
    }

    public async Task SendCredentialErrorNotificationAsync(string serviceName, string message, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrEmpty(_webhookUrl)) return;
        var (content, contentType) = FormatAlertPayload("Credential error", $"{serviceName}: {message}");
        var client = _httpClientFactory.CreateClient();
        var body = new StringContent(content, Encoding.UTF8, contentType);
        await client.PostAsync(_webhookUrl, body, cancellationToken).ConfigureAwait(false);
    }

    public async Task TestWebhookAsync(CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrEmpty(_webhookUrl)) return;
        var (content, contentType) = FormatAlertPayload("Test", "Asset Sync webhook test.");
        var client = _httpClientFactory.CreateClient();
        var body = new StringContent(content, Encoding.UTF8, contentType);
        await client.PostAsync(_webhookUrl, body, cancellationToken).ConfigureAwait(false);
    }

    private (string Content, string ContentType) FormatSyncPayload(SyncRunSummary summary)
    {
        if (_webhookType == "Teams")
        {
            var card = new
            {
                type = "message",
                attachments = new[]
                {
                    new
                    {
                        contentType = "application/vnd.microsoft.card.adaptive",
                        content = new
                        {
                            type = "AdaptiveCard",
                            version = "1.0",
                            body = new object[]
                            {
                                new { type = "TextBlock", text = "Asset Sync completed", weight = "bolder", size = "large" },
                                new { type = "FactSet", facts = new[] { new { title = "Created", value = summary.Created.ToString() }, new { title = "Updated", value = summary.Updated.ToString() }, new { title = "Skipped", value = summary.Skipped.ToString() }, new { title = "Errors", value = summary.Errors.ToString() } } }
                            }
                        }
                    }
                }
            };
            return (JsonSerializer.Serialize(card), "application/json");
        }
        if (_webhookType == "Slack")
        {
            var block = new { text = new { type = "mrkdwn", text = $"*Asset Sync* completed. Created: {summary.Created}, Updated: {summary.Updated}, Skipped: {summary.Skipped}, Errors: {summary.Errors}" } };
            return (JsonSerializer.Serialize(new { blocks = new[] { block } }), "application/json");
        }
        return (JsonSerializer.Serialize(summary), "application/json");
    }

    private (string Content, string ContentType) FormatAlertPayload(string title, string message)
    {
        if (_webhookType == "Teams")
        {
            var card = new { type = "message", attachments = new[] { new { contentType = "application/vnd.microsoft.card.adaptive", content = new { type = "AdaptiveCard", version = "1.0", body = new object[] { new { type = "TextBlock", text = title, weight = "bolder" }, new { type = "TextBlock", text = message } } } } } };
            return (JsonSerializer.Serialize(card), "application/json");
        }
        if (_webhookType == "Slack")
            return (JsonSerializer.Serialize(new { text = $"{title}: {message}" }), "application/json");
        return (JsonSerializer.Serialize(new { title, message }), "application/json");
    }
}

File: AssetSync-macOS/AssetSyncInfrastructure/Sources/AssetSyncInfrastructure/Api/WebhookService.swift
import Foundation
import AssetSyncCore

/// Sends notifications to Teams, Slack, or generic webhook endpoints.
public final class WebhookService: WebhookServiceProtocol, @unchecked Sendable {
    private let webhookURL: String?
    private let webhookType: String
    private let session: URLSession

    public init(webhookURL: String?, webhookType: String = "Generic", session: URLSession = .shared) {
        self.webhookURL = webhookURL
        self.webhookType = webhookType
        self.session = session
    }

    public func sendSyncNotification(_ summary: SyncRunSummary) async {
        guard let url = webhookURL, !url.isEmpty else { return }
        let payload = formatSyncPayload(summary)
        await post(to: url, json: payload)
    }

    public func sendConnectivityFailureNotification(serviceName: String, message: String) async {
        guard let url = webhookURL, !url.isEmpty else { return }
        let payload = formatAlertPayload(title: "Connectivity failure", message: "\(serviceName): \(message)")
        await post(to: url, json: payload)
    }

    public func sendCredentialErrorNotification(serviceName: String, message: String) async {
        guard let url = webhookURL, !url.isEmpty else { return }
        let payload = formatAlertPayload(title: "Credential error", message: "\(serviceName): \(message)")
        await post(to: url, json: payload)
    }

    public func testWebhook() async {
        guard let url = webhookURL, !url.isEmpty else { return }
        let payload = formatAlertPayload(title: "Test", message: "Asset Sync webhook test.")
        await post(to: url, json: payload)
    }

    // MARK: - Formatting

    private func formatSyncPayload(_ summary: SyncRunSummary) -> [String: Any] {
        switch webhookType {
        case "Teams":
            return [
                "type": "message",
                "attachments": [[
                    "contentType": "application/vnd.microsoft.card.adaptive",
                    "content": [
                        "type": "AdaptiveCard", "version": "1.0",
                        "body": [
                            ["type": "TextBlock", "text": "Asset Sync completed", "weight": "bolder", "size": "large"],
                            ["type": "FactSet", "facts": [
                                ["title": "Created", "value": "\(summary.created)"],
                                ["title": "Updated", "value": "\(summary.updated)"],
                                ["title": "Skipped", "value": "\(summary.skipped)"],
                                ["title": "Errors", "value": "\(summary.errors)"],
                            ]],
                        ],
                    ] as [String : Any],
                ]],
            ]
        case "Slack":
            return [
                "blocks": [[
                    "text": [
                        "type": "mrkdwn",
                        "text": "*Asset Sync* completed. Created: \(summary.created), Updated: \(summary.updated), Skipped: \(summary.skipped), Errors: \(summary.errors)",
                    ],
                ]],
            ]
        default:
            return [
                "created": summary.created, "updated": summary.updated,
                "skipped": summary.skipped, "errors": summary.errors,
                "dryRun": summary.dryRun, "syncRunId": summary.syncRunId,
            ]
        }
    }

    private func formatAlertPayload(title: String, message: String) -> [String: Any] {
        switch webhookType {
        case "Teams":
            return [
                "type": "message",
                "attachments": [[
                    "contentType": "application/vnd.microsoft.card.adaptive",
                    "content": ["type": "AdaptiveCard", "version": "1.0", "body": [
                        ["type": "TextBlock", "text": title, "weight": "bolder"],
                        ["type": "TextBlock", "text": message],
                    ]] as [String : Any],
                ]],
            ]
        case "Slack":
            return ["text": "\(title): \(message)"]
        default:
            return ["title": title, "message": message]
        }
    }

    private func post(to urlString: String, json: [String: Any]) async {
        guard let url = URL(string: urlString) else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: json)
        _ = try? await session.data(for: request)
    }
}
